brpc.cpp

```diff
@@ -14,18 +14,118 @@ CHECK_MEMORY_LEAKS
 #include "src/type/ObjectMap.h"
 #include "src/rpc/server/RpcServer.h"
 
+#ifndef WIN32
+#include <signal.h>
+#include <pthread.h>
+
+class SignalHandler
+{
+public:
+	static void catchSignals(int sigs[])
+	{
+		//signal(SIGSEGV, signalHandle);
+		for(int i = 0; sigs[i] != 0; i++)
+		{
+			assert(SIG_ERR != signal(sigs[i], signalHandle));
+		}
+	}
+
+	static void signalHandle(int sig)
+	{
+		printf("Signal %d\n", sig);
+		Util::dumpStack();
+		exit(-1);
+	}
+
+	static void ignoreSignal(int sig) throw(Exception)
+	{
+		struct sigaction sa;
+		sa.sa_handler = SIG_IGN;
+		sa.sa_flags = 0;
+		if(sigemptyset(&sa.sa_mask) == -1 || sigaction(sig, &sa, 0) == -1)
+		{
+			throw Exception(String::format(
+					"Failed to set ignore signal %d.", sig));
+		}
+
+		sigset_t signal_mask;
+		sigemptyset(&signal_mask);
+		sigaddset(&signal_mask, sig);
+		if(pthread_sigmask(SIG_BLOCK, &signal_mask, NULL) != 0)
+		{
+			throw Exception(String::format(
+					"Failed to set pthread ignore signal %d.", sig));
+		}
+	}
+
+	static void ignoreSignals(int sigs[]) throw(Exception)
+	{
+		for(int i = 0; sigs[i] != 0; i++)
+		{
+			ignoreSignal(sigs[i]);
+		}
+	}
+};
+
+// for linux like
+static void initSignalHandler() throw(Exception)
+{
+	int catches[] = {SIGSEGV, SIGBUS, SIGABRT, 0};
+	SignalHandler::catchSignals(catches);
+
+	int ignores[] = {SIGPIPE, 0};
```

Comment:
ignore signal SIGPIPE to prevent the program from crashing when the socket is interrupted.